arithmetization/src/main/java/net/consensys/linea/sequencer/txselection/selectors/ProfitableTransactionSelector.java

```diff
@@ -30,11 +40,80 @@
 @Slf4j
 @RequiredArgsConstructor
 public class ProfitableTransactionSelector implements PluginTransactionSelector {
+  @VisibleForTesting protected static Set<Hash> unprofitableCache = new LinkedHashSet<>();
+  @VisibleForTesting protected static Wei prevMinGasPrice = Wei.MAX_WEI;
+
   private final int verificationGasCost;
   private final int verificationCapacity;
   private final int gasPriceRatio;
   private final double minMargin;
   private final int adjustTxSize;
+  private final int unprofitableCacheSize;
+  private final int unprofitableRetryLimit;
+  private int unprofitableRetries;
+  private MutableBoolean minGasPriceDecreased;
+
+  @Override
+  public TransactionSelectionResult evaluateTransactionPreProcessing(
+      final TransactionEvaluationContext<? extends PendingTransaction> evaluationContext) {
+
+    final Wei minGasPrice = evaluationContext.getMinGasPrice();
+
+    // update prev min gas price only if it is a new block
+    if (minGasPriceDecreased == null) {
+      minGasPriceDecreased = new MutableBoolean(minGasPrice.lessThan(prevMinGasPrice));
+      prevMinGasPrice = minGasPrice;
+    }
+
+    if (!evaluationContext.getPendingTransaction().hasPriority()) {
+      final Transaction transaction = evaluationContext.getPendingTransaction().getTransaction();
+      final double effectiveGasPrice =
+          evaluationContext.getTransactionGasPrice().getAsBigInteger().doubleValue();
+      final long gasLimit = transaction.getGasLimit();
+
+      // check the upfront profitability using the gas limit of the tx
+      if (!isProfitable(
+          "PreProcessing",
+          transaction,
+          minGasPrice.getAsBigInteger().doubleValue(),
+          effectiveGasPrice,
+          gasLimit)) {
+        return TX_UNPROFITABLE_UPFRONT;
+      }
+
+      if (unprofitableCache.contains(transaction.getHash())) {
+        // only retry unprofitable txs if the min gas price went down
+        if (minGasPriceDecreased.isTrue()) {
```

Comment:
This is useful to tell when we definitely don't want to reprocess any unprofitable transactions (during gas price increase), but it's not very useful for cases when gas price fluctuates. I think we can improve this by storing gas price at which given tx was considered unprofitable and take this into account.

Note that `PriorityBlockingQueue` sorted by checked gas price won't work well here because there is no monotonous dependency between profitability and gas price. Heap sorted by profitability may work though